File: typescript/src/journal.ts
/**
 * Journal — durable execution primitive.
 *
 * Append-only WAL on disk. Every step is recorded with its result.
 * On crash, replayed steps return cached results instead of re-executing.
 */

import { createHash } from "crypto";
import { existsSync, mkdirSync, readFileSync, appendFileSync, unlinkSync, writeFileSync } from "fs";
import { join } from "path";

interface JournalEntry {
  step: number;
  tag: string;
  argsHash: string;
  result: unknown;
  ts: number;
}

export class Journal {
  private dir: string;
  private walPath: string;
  private entries = new Map<string, JournalEntry>();
  private stepCounter = 0;

  constructor(path: string) {
    this.dir = path;
    this.walPath = join(path, "wal.jsonl");
    mkdirSync(path, { recursive: true });
    this.load();
  }

  private load(): void {
    if (!existsSync(this.walPath)) return;
    const lines = readFileSync(this.walPath, "utf-8").split("\n");
    for (const line of lines) {
      if (!line.trim()) continue;
      const data = JSON.parse(line);
      const key = `${data.step}:${data.tag}:${data.args_hash}`;
      this.entries.set(key, {
        step: data.step,
        tag: data.tag,
        argsHash: data.args_hash,
        result: data.result,
        ts: data.ts,
      });
      this.stepCounter = Math.max(this.stepCounter, data.step + 1);
    }
  }

  async durable<T>(tag: string, fn: () => Promise<T>, ...args: unknown[]): Promise<T> {
    const step = this.stepCounter++;
    const argsHash = hashArgs(args);
    const key = `${step}:${tag}:${argsHash}`;

    const cached = this.entries.get(key);
    if (cached !== undefined) {
      return cached.result as T;
    }

    const result = await fn();
    const entry: JournalEntry = { step, tag, argsHash, result, ts: Date.now() / 1000 };
    this.entries.set(key, entry);

    const record = JSON.stringify({
      step,
      tag,
      args_hash: argsHash,
      result,
      ts: entry.ts,
    });
    appendFileSync(this.walPath, record + "\n");

    return result;
  }

  reset(): void {
    this.entries.clear();
    this.stepCounter = 0;
    if (existsSync(this.walPath)) unlinkSync(this.walPath);
  }

  get entryCount(): number {
    return this.entries.size;
  }
}

function hashArgs(args: unknown[]): string {
  const raw = JSON.stringify(args);
  return createHash("sha256").update(raw).digest("hex").slice(0, 16);
}

File: python/oncell/journal.py
"""Journal — durable execution primitive.

Append-only WAL on NVMe. Every step is recorded with its result.
On crash, replayed steps return cached results instead of re-executing.

Usage (automatic via cell.shell with durable=True):
    result = await cell.shell("npm test")  # journaled by default

Usage (manual for custom durability):
    async with cell.journal.step("fetch_data") as step:
        if step.cached:
            data = step.result
        else:
            data = await fetch_data()
            step.record(data)
"""

from __future__ import annotations

import hashlib
import json
import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable


@dataclass
class JournalEntry:
    step: int
    tag: str
    args_hash: str
    result: Any
    timestamp: float


class Journal:
    def __init__(self, path: str | Path):
        self._path = Path(path)
        self._path.mkdir(parents=True, exist_ok=True)
        self._wal = self._path / "wal.jsonl"
        self._entries: dict[str, JournalEntry] = {}
        self._step_counter = 0
        self._load()

    def _load(self):
        if not self._wal.exists():
            return
        with open(self._wal, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                data = json.loads(line)
                key = f"{data['step']}:{data['tag']}:{data['args_hash']}"
                self._entries[key] = JournalEntry(
                    step=data["step"],
                    tag=data["tag"],
                    args_hash=data["args_hash"],
                    result=data["result"],
                    timestamp=data["ts"],
                )
                self._step_counter = max(self._step_counter, data["step"] + 1)

    def _key(self, step: int, tag: str, args_hash: str) -> str:
        return f"{step}:{tag}:{args_hash}"

    async def durable(self, tag: str, fn: Callable, *args: Any, **kwargs: Any) -> Any:
        """Execute fn with durable journaling. Returns cached result on replay."""
        step = self._step_counter
        self._step_counter += 1
        args_hash = _hash_args(*args, **kwargs)
        key = self._key(step, tag, args_hash)

        cached = self._entries.get(key)
        if cached is not None:
            return _deserialize(tag, cached.result)

        result = await fn()
        serialized = _serialize(result)

        entry = JournalEntry(step=step, tag=tag, args_hash=args_hash, result=serialized, timestamp=time.time())
        self._entries[key] = entry

        record = {"step": step, "tag": tag, "args_hash": args_hash, "result": serialized, "ts": entry.timestamp}
        with open(self._wal, "a") as f:
            f.write(json.dumps(record, default=str) + "\n")
            f.flush()
            os.fsync(f.fileno())

        return result

    def reset(self):
        """Clear the journal."""
        self._entries.clear()
        self._step_counter = 0
        if self._wal.exists():
            self._wal.unlink()

    @property
    def entries(self) -> int:
        return len(self._entries)


def _hash_args(*args: Any, **kwargs: Any) -> str:
    raw = json.dumps({"a": args, "k": kwargs}, sort_keys=True, default=str)
    return hashlib.sha256(raw.encode()).hexdigest()[:16]


def _serialize(result: Any) -> Any:
    """Convert to JSON-safe form."""
    if hasattr(result, "__dataclass_fields__"):
        from dataclasses import asdict
        return asdict(result)
    return result


def _deserialize(tag: str, data: Any) -> Any:
    """Reconstruct typed result from journal data."""
    if tag == "shell" and isinstance(data, dict) and "exit_code" in data:
        from oncell.cell import ShellResult
        return ShellResult(**data)
    return data
